amlb/results.py

```diff
@@ -169,7 +169,7 @@ def as_printable_data_frame(self, verbosity=3):
         for col in nanable_int_cols:
             df[col] = df[col].astype(np.object).map(int_print).astype(np.str)
         for col in low_precision_float_cols:
-            df[col] = df[col].astype(np.float).map(partial(num_print, "{:.1f}".format)).astype(np.float)
+            df[col] = df[col].astype(np.float).map(partial(num_print, "{:.3g}".format)).astype(np.float)
```

Comment:
hold on, will come up with a better formatting:
```python
> "{:0.3g}".format(12345)
'1.23e+04'
```
which I don't like, I think we lose readability if all numbers are not printed the same way
`g` is great for small numbers but `f` is better for large ones…

Reply:
Also, for those duration measurements, I don't think we can guarantee a precision better than 0.1s (and it's already very optimistic given that some frameworks get the data loaded before the timer, and others during…).
I think the main goal here is just to avoid having `0` printed for very low values, like some prediction times on some very small datasets (for example with `flaml`).

Reply:
I think when visualizing the results we have to separate packages which load data internally and those which get passed a pandas object.

Reply:
```suggestion
            float_format = lambda f: ("{:.1g}" if f < 1 else "{:.1f}").format(f)
            df[col] = df[col].astype(np.float).map(partial(num_print, float_format)).astype(np.float)
```

Reply:
@PGijsbers it's not only a question of data loading: as soon as we provide a measure, users will blindly do some ranking on it, and  due to the inherent imprecision in time measurement for such a complex feature, if for example you have 2 predict durations with `20.311` and `20.317`, I don't think it's fair to rank one higher than the other.
I don't see any way to completely prevent this, but I think we should at least mitigate those readings by trying to offer meaningful numbers.

Reply:
But that's also why `g` might be good, since it would be `20.3` if only 3 significant digits are shown both times. That said, I understand that there's also a minimum measurable duration, and I don't really know what that is in our application. Normally measuring with milliseconds is OK (I think?), though I'm not 100% sure if that works well enough here (and such short tests are normally repeated, whereas here it's a one-off).

Reply:
As said before, we're mainly concerned with not displaying `0.0` for small prediction times which occur *very rarely*: (too) small datasets + simple models (e.g. flaml), so for the great majority of durations measured `f` is perfectly fine as it renders *all* digits with the requested precision (0.1). If you use `g` everywhere then by limiting the number of significant digits, the precision also varies depending on how large the number is: this is fine for a metric like `mse` but for a duration, the precision is fixed, it just depends on how the timer is applied. 

On top of this, `g` will use scientific notation for most durations, which doesn't make it friendly when looking directly at the results in the console or in the csv file. I do that constantly, and for me `3600` = 1h instantly, but `3.6e+03`…takes about a second more, not even mentioning `1.44e+04`…

Now the question of precision. Sure, the timer is accurate at least to the millisecond, but what we're timing is not.
> though I'm not 100% sure if that works well enough here (and such short tests are normally repeated, whereas here it's a one-off)

Exactly, microbenchmarks for example rely on a large amount of repetitions to attain this level of accuracy.
In our case, on top of this, the way the timer is applied is not fully consistent (check the framework integrations): some frameworks systematically provide the probabilities, some don't. For sklearn-based APIs, I was at first timing only the call to `predict`, but some frameworks changed this to measure `predict_proba`, a framework like `MLPlan` returns its own prediction time whereas at the opposite for `MLNet` we measure the lifetime of an entire subprocess… that's why I'm saying that providing a `.1` precision is already generous, and I'm not comfortable pretending we can provide millis.

However, accuracy is improved by considering prediction per row when analyzing the results: the larger the dataset, the more accurate the `prediction_duration/nrows`, and it allows us to neglect some of those integration differences mentioned above. If you have a precision of .1s for the entire prediction duration, then for 100 rows, you already have a precision around millis.

With my suggestion above, all numbers are rendered "plain" with the same precision, except for the very small ones for which we're just avoiding the `0.0`:
```python
In [1]: float_format = lambda f: ("{:.1g}" if f < 1 else "{:.1f}").format(f)

In [2]: float_format(3634.12345)
Out[2]: '3634.1'

In [3]: float_format(14489.3456)
Out[3]: '14489.3'

In [4]: float_format(123.987)
Out[4]: '124.0'

In [5]: float_format(0.5678)
Out[5]: '0.6'

In [6]: float_format(0.00345)
Out[6]: '0.003'
```

Finally right now, when we look at prediction time, `.1` already allows us to clearly separate frameworks, so I don't see the need to add more digits that we can't justify clearly.